### PR TITLE
Add token generator for offline pages

### PR DESCRIPTION
**Pull request by AidanCurrah**
This PR add a token generator that accepts a `path` and returns a hash. In addition, the `OnlineBase` model has a new function `get_preview_url` that appends `?preview=PATH_HASH` if the model is offline. Finally, the `PublicationMiddleware` now has two different conditions for allowing a user to view an offline page.
1. The same as the old check, except it just checks if the GET parameter `preview` exists
2. The value of the GET parameter `preview` is equal to the hash of the requested `path`.

This means elsewhere (OSM Jet), we can add links to `object.get_preview_url` to have shareable links to offline pages that can be viewed by people not logged in.

### CODE REVIEW

File: cms/models/base.py [hunk 1]
```diff
@@ -1,13 +1,36 @@
...
 from watson.search import SearchAdapter
 
 from cms.apps.media.models import ImageRefField
 from cms.models.managers import (OnlineBaseManager, PageBaseManager,
                                  PublishedBaseManager, SearchMetaBaseManager)
 
 
+class PathTokenGenerator:
+    """
+    Strategy object used to generate and check tokens for the password
```
Comment (lewiscollard) [inline cms/models/base.py]: not a password reset mechanism :)

File: cms/models/base.py [hunk 2]
```diff
@@ -40,6 +63,15 @@ class OnlineBase(PublishedBase):
         ),
     )
 
+    def get_preview_url(self):
+        if not hasattr(self, 'get_absolute_url'):
+            return None
+
+        if self.is_online:
```
Comment (lewiscollard) [inline cms/models/base.py]: This won't work for anything else that subclasses `OnlineBaseManager` or `PageBaseManager` and implements extra publication controls, such as the Page system having a publication and expiry date, or (for our standard news app) a date in the future, or any number of other possible cases.

You can't reasonably catch this without a massive abstraction inversion - it's fine to return the preview URL unconditionally here.

**Comment by lewiscollard** [comment]
I actually really like the simplicity of this approach. :)

File: cms/middleware.py [hunk 3]
```diff
@@ -37,11 +37,13 @@ def process_request(self, request):
         if not any(pattern.match(request.path_info[1:]) for pattern in exclude_urls):
             # See if preview mode is requested.
             try:
-                preview_mode = bool(int(request.GET.get('preview', 0)))
+                path = f'{request.path_info[1:] if request.path_info[1:] else request.path_info}'
+                token_preview_valid = path_token_generator.check_token(request.GET.get('preview', 0), path)
+                user_preview = request.GET.get('preview', 0) and request.user.is_authenticated() and request.user.is_staff and request.user.is_active
```
Comment (daniel-clayton) [inline cms/middleware.py]: You don't need to check `is_active` if they are authenticated

File: cms/middleware.py [hunk 4]
```diff
@@ -37,11 +37,13 @@ def process_request(self, request):
         if not any(pattern.match(request.path_info[1:]) for pattern in exclude_urls):
             # See if preview mode is requested.
             try:
-                preview_mode = bool(int(request.GET.get('preview', 0)))
+                path = f'{request.path_info[1:] if request.path_info[1:] else request.path_info}'
+                token_preview_valid = path_token_generator.check_token(request.GET.get('preview', 0), path)
+                user_preview = request.GET.get('preview', 0) and request.user.is_authenticated() and request.user.is_staff
```
Comment (lewiscollard) [inline cms/middleware.py]: Change `is_authenticated()` to `is_authenticated` while you are here - the callable version was deprecated in 1.11 and removed in 2.x.

**Pull request merged**